sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,748 @@
+      }
+      channel.txCommit();
+      oldestTimestamp = Instant.now();
+      sessionIds.clear();
+    }
+
+  }
+
+  private static class UnboundedRabbitMqReader extends UnboundedSource.UnboundedReader<byte[]> {
+
+    private final RabbitMQSource source;
+
+    private byte[] current;
+    private byte[] currentRecordId;
+    private Connection connection;
+    private Channel channel;
+    private QueueingConsumer consumer;
+    private Instant currentTimestamp;
+    private RabbitMQCheckpointMark checkpointMark;
+
+    public UnboundedRabbitMqReader(RabbitMQSource source,
+                                   RabbitMQCheckpointMark checkpointMark) {
+      this.source = source;
+      this.current = null;
+      if (checkpointMark != null) {
+        this.checkpointMark = checkpointMark;
+      } else {
+        this.checkpointMark = new RabbitMQCheckpointMark();
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return checkpointMark.oldestTimestamp;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public RabbitMQSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public byte[] getCurrentRecordId() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      if (currentRecordId != null) {
+        return currentRecordId;
+      } else {
+        return "".getBytes();
+      }
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (currentTimestamp == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory;
+      try {
+        connectionFactory =
+            source.spec.connectionConfig().createConnectionFactory();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      try {
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+        String queueName = null;
+        if (source.spec.connectionConfig().queue() != null) {
+          channel.queueDeclare(source.spec.connectionConfig().queue(), false, false, false, null);
+          queueName = source.spec.connectionConfig().queue();
+        } else if (source.spec.connectionConfig().exchange() != null) {
+          channel.exchangeDeclare(source.spec.connectionConfig().exchange(),
+              source.spec.connectionConfig().exchangeType());
+          queueName = channel.queueDeclare().getQueue();
+          channel.queueBind(queueName, source.spec.connectionConfig().exchange(), "");
```

Comment:
Feels like it'd be useful to allow specifying the binding key as well, esp. for some exchange types.

Reply:
Agree, I will add an optional `withBindingKey()` and maybe `withRoutingKey()` as well.